### ISSUE TITLE
util_helpers.c fails to build on ARM due to "{standard input}:39: Error: ARM register expected -- `ldrb r4,[$1] ldrb r5,[$2] " [...]

### ISSUE THREAD

**Issue by jmesmon** [DaGenix/rust-crypto#302]
```
| Process didn't exit successfully: `/home/cody/obj/y/tmp/work/cortexa8hf-vfp-neon-poky-linux-gnueabi/coattail/0.0.0+gitAUTOINC+7882984c93-r0/git/target/release/build/rust-crypto-67f1ad94f3ffe5f4/build-script-build` (exit code: 101)
| --- stdout
| TARGET = Some("arm-poky-linux-gnueabi")
| TARGET = Some("arm-poky-linux-gnueabi")
| CARGO_MANIFEST_DIR = Some("/home/cody/obj/y/tmp/work/cortexa8hf-vfp-neon-poky-linux-gnueabi/coattail/0.0.0+gitAUTOINC+7882984c93-r0/cargo_home/registry/src/github.com-1ecc6299db9ec823/rust-crypto-0.2.31")
| OUT_DIR = Some("/home/cody/obj/y/tmp/work/cortexa8hf-vfp-neon-poky-linux-gnueabi/coattail/0.0.0+gitAUTOINC+7882984c93-r0/git/target/arm-poky-linux-gnueabi/release/build/rust-crypto-67f1ad94f3ffe5f4/out")
| OPT_LEVEL = Some("3")
| PROFILE = Some("release")
| release 3
| TARGET = Some("arm-poky-linux-gnueabi")
| HOST = Some("x86_64-linux")
| CC_arm-poky-linux-gnueabi = None
| CC_arm_poky_linux_gnueabi = None
| TARGET_CC = Some("arm-poky-linux-gnueabi-gcc")
| TARGET = Some("arm-poky-linux-gnueabi")
| HOST = Some("x86_64-linux")
| CFLAGS_arm-poky-linux-gnueabi = None
| CFLAGS_arm_poky_linux_gnueabi = None
| TARGET_CFLAGS = Some(" -march=armv7-a -mfloat-abi=hard -mfpu=neon -mtune=cortex-a8 --sysroot=/home/cody/obj/y/tmp/sysroots/beaglebone  -O2 -pipe -g -feliminate-unused-debug-types")
| TARGET = Some("arm-poky-linux-gnueabi")
| HOST = Some("x86_64-linux")
| CC_arm-poky-linux-gnueabi = None
| CC_arm_poky_linux_gnueabi = None
| TARGET_CC = Some("arm-poky-linux-gnueabi-gcc")
| running: "arm-poky-linux-gnueabi-gcc" "-O3" "-c" "-ffunction-sections" "-fdata-sections" "-march=armv7-a" "-mfloat-abi=hard" "-mfpu=neon" "-mtune=cortex-a8" "--sysroot=/home/cody/obj/y/tmp/sysroots/beaglebone" "-O2" "-pipe" "-g" "-feliminate-unused-debug-types" "-fPIC" "/home/cody/obj/y/tmp/work/cortexa8hf-vfp-neon-poky-linux-gnueabi/coattail/0.0.0+gitAUTOINC+7882984c93-r0/cargo_home/registry/src/github.com-1ecc6299db9ec823/rust-crypto-0.2.31/src/util_helpers.c" "-o" "/home/cody/obj/y/tmp/work/cortexa8hf-vfp-neon-poky-linux-gnueabi/coattail/0.0.0+gitAUTOINC+7882984c93-r0/git/target/arm-poky-linux-gnueabi/release/build/rust-crypto-67f1ad94f3ffe5f4/out/src/util_helpers.o"
| 
| 
| command did not execute successfully, got: exit code: 1
| 
| 
| 
| --- stderr
| {standard input}: Assembler messages:
| {standard input}:39: Error: ARM register expected -- `ldrb r4,[$1] ldrb r5,[$2] eor r4,r4,r5 orr $0,$0,r4 add $1,$1,#1 add $2,$2,#1 subs $3,$3,#1 bne 1b'
| thread '<main>' panicked at 'explicit panic', /home/cody/obj/y/tmp/work/cortexa8hf-vfp-neon-poky-linux-gnueabi/coattail/0.0.0+gitAUTOINC+7882984c93-r0/cargo_home/registry/src/github.com-1ecc6299db9ec823/gcc-0.3.6/src/lib.rs:459
| stack backtrace:
|    1:     0x7fe9986b8458 - sys::backtrace::write::hd8c95f651560dfe4y6r
|    2:     0x7fe9986c1244 - panicking::on_panic::h441ceaa2b70ecb64JMw
|    3:     0x7fe99867d6e2 - rt::unwind::begin_unwind_inner::h080f1ff0e70a38bcTrw
|    4:           0x409c20 - rt::unwind::begin_unwind::h17406813340456516896
|    5:           0x40a78c - run::haf0d0b3462d18e1fBva
|    6:           0x40725a - Config::compile::ha709ea40cfe7f8f3Wga
|    7:           0x4040b5 - compile_library::h5248a1a1b8319f602ca
|    8:           0x4034e5 - main::h6ac4eb617961c7e9faa
|                         at /home/cody/obj/y/tmp/work/cortexa8hf-vfp-neon-poky-linux-gnueabi/coattail/0.0.0+gitAUTOINC+7882984c93-r0/cargo_home/registry/src/github.com-1ecc6299db9ec823/rust-crypto-0.2.31/build.rs:10
|    9:     0x7fe998745a78 - rust_try_inner
|   10:     0x7fe998745a65 - rust_try
|   11:     0x7fe9986c2d53 - rt::lang_start::h11cfe57bca456a8deHw
|   12:     0x7fe997e1278f - __libc_start_main
|   13:           0x4033a8 - _start
|   14:                0x0 - <unknown>
| 
```

**Comment by jmesmon** [comment]
I'd consider using a more portable implimentation of a timing safe memcmp, like:

 - openbsd's: http://cvsweb.openbsd.org/cgi-bin/cvsweb/~checkout~/src/lib/libc/string/timingsafe_memcmp.c?rev=1.1&content-type=text/plain
 - openssl's: https://github.com/openssl/openssl/blob/master/crypto/cryptlib.c#L459-L470
 - gnutls's: https://gitlab.com/gnutls/gnutls/blob/master/lib/safe-memfuncs.c#L70-82 (very similar to openssls)

All of these are in plain C.

**Comment by petevine** [comment]
I got the same error trying to build coreutils:

https://github.com/uutils/coreutils/issues/701

**Comment by friedm** [comment]
I got this today when building for armv7 on a raspberry pi.

**Comment by DaGenix** [comment]
Thanks for the report and sorry for taking so long to get to it. Yikes - it looks like I didn't test the ARM build before pushing that code out: it was missing semicolons between statements and was using the wrong syntax to name register arguments. Basically, it was copy-pasted from when it used the Rust `asm!()` macro where both of those things are not an error. Anyway, I think this is fixed by #321. Please let me know if you are still having problems.

The problem with doing this in C / Rust is that there really isn't any way of guaranteeing that a sufficiently smart compiler optimization won't re-write the method to short circuit the comparison as soon as a difference is found, breaking the function's timing-attack resistance. You can look at the generated assembly to make sure that its timing-attack resistant, but, you have to do that for every LLVM / Gcc version. By writing it in assembly, you can be sure that the compiler won't mess with it. It might be possible to do all the work in C / Rust code and then feed the result into an empty `asm` block in order to prevent those some of those types of optimizations. However, I'm not sure that it will prevent every possible optimization that could break the timing-attack resistance.

**Comment by petevine** [comment]
Yep, build just fine on arm, thanks!
